### PR TITLE
[MM-9374] Open Permalinks in app

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Open permalinks in-app the same way we do the search result preview

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-9374

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] All new/modified APIs include changes to [mattermost-redux](https://github.com/mattermost/mattermost-redux) (https://github.com/mattermost/mattermost-redux/pull/408)
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: 
* iOS 11.2 iPhone 6 and iPhone X Simulator
* 11.2.5 iPhone 6s
* Android 7.1.1 and Android 8.0 Emulator
* Android 6.0.1 Samsung J5

### CODE REVIEW

File: app/components/markdown/markdown_link/markdown_link.js [hunk 1]
```diff
@@ -12,36 +12,58 @@ import mattermostManaged from 'app/mattermost_managed';
...
         }
 
-        Linking.canOpenURL(url).then((supported) => {
-            if (supported) {
-                Linking.openURL(url);
-            }
-        });
-    };
+        const pattern = new RegExp('^' + escapeRegex(serverURL) + '\\/([^\\/]+)\\/pl\\/(\\w+)');
+        const sitePattern = new RegExp('^' + escapeRegex(siteURL) + '\\/([^\\/]+)\\/pl\\/(\\w+)');
```
Comment (saturninoabril) [inline app/components/markdown/markdown_link/markdown_link.js]: Very minor and is non-blocking: just thinking if it makes sense to have the entire line be part of `escapeRegex` function?
```
new RegExp('^' + escapeRegex(siteURL) + '\\/([^\\/]+)\\/pl\\/(\\w+)');
``

Comment (hmhealey) [reply app/components/markdown/markdown_link/markdown_link.js]: The `escapeRegex` function only matters for the serverURL and siteURL since those are user defined, and having characters like `*` or `[` in them would cause the `new RegExp` to error out. The rest of the pattern is already escaped.

File: app/components/markdown/markdown_link/markdown_link.js [hunk 2]
```diff
@@ -12,36 +12,58 @@ import mattermostManaged from 'app/mattermost_managed';
...
+                if (supported) {
+                    Linking.openURL(url);
+                }
+            });
+            return;
+        }
+
+        const teamName = match[1];
+        const postId = match[2];
+        actions.loadChannelsByTeamName(teamName);
```
Comment (hmhealey) [inline app/components/markdown/markdown_link/markdown_link.js]: Why is this done here? Shouldn't it be done wherever `onPermalinkPress` is defined or by the permalink view itself?

File: app/components/post_list/post_list.js [hunk 3]
```diff
@@ -96,25 +99,64 @@ export default class PostList extends PureComponent {
...
     }
 
+    handleClosePermalink = () => {
+        const {actions} = this.props;
+        actions.selectFocusedPostId('');
+        this.setState({showingPermalink: false});
+    };
+
+    handlePermalinkPress = (postId) => {
+        this.setState({isPermalink: true});
```
Comment (hmhealey) [inline app/components/post_list/post_list.js]: Is `this.state.isPermalink` needed? It looks like we set it here and then immediately reference it below which may not have the value you've set here if I understand React's state updating properly. If it does have the new value of `this.state.isPermalink` below, then we're always passing `isPermalink: true`  into the Permalink screen from here, so storing it in state is unnecessary.

If we do end up needing it in state, we should define an initial value since `this.state.isPermalink` is initially undefined

File: app/components/post_list/post_list.js [hunk 4]
```diff
@@ -96,25 +99,64 @@ export default class PostList extends PureComponent {
...
+        this.showPermalinkView(postId);
+    };
+
+    showPermalinkView = (postId) => {
+        const {actions, navigator} = this.props;
+        const {isPermalink, showingPermalink} = this.state;
```
Comment (hmhealey) [inline app/components/post_list/post_list.js]: Maybe move `this.state.showingPermalink` to be just `this.showingPermalink` in case we get caught by state not being updated in time?

File: app/screens/channel_info/channel_info.js [hunk 5]
```diff
@@ -234,6 +241,45 @@ class ChannelInfo extends PureComponent {
...
+    };
+
+    handlePermalinkPress = (postId) => {
+        this.setState({isPermalink: true});
+        this.showPermalinkView(postId);
+    };
+
+    showPermalinkView = (postId) => {
+        const {actions, navigator} = this.props;
+        const {isPermalink, showingPermalink} = this.state;
```
Comment (hmhealey) [inline app/screens/channel_info/channel_info.js]: Same comments as above about `isPermalink` seeming unnecessary and `showingPermalink` being more reliable outside of state

File: app/screens/permalink/permalink.js [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+
+        const {postIds, channelName} = props;
+        let show = false;
+
+        if (postIds && postIds.length >= 10) {
+            show = true;
+        }
+
+        if (!show) {
+            this.loadPosts(props, context);
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: Could you move this to `componentDidMount` or `componentWillMount`? That's where we usually do any required loading

File: app/screens/permalink/permalink.js [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+        }
+
+        this.state = {
+            title: channelName,
+            show,
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: This sounds a bit like this controls whether or not the whole screen is visible. Maybe something like `loading` would be more clear? 0/5 though

File: app/screens/permalink/permalink.js [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+        };
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (this.props.channelName !== nextProps.channelName) {
+            this.setState({title: nextProps.channelName});
+        }
+
+        if (this.props.focusedPostId !== nextProps.focusedPostId) {
+            this.setState({show: false}, () => {
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: Maybe we should only change `this.state.show` if `nextProps.postIds && nextProps.postIds.length < 10`? I don't follow why we'd show this as loading when changing focused posts if we don't need to

File: app/screens/permalink/permalink.js [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+        };
+
+        navigator.push(options);
+    });
+
+    handleClose = () => {
+        const {actions, navigator, onClose} = this.props;
+        if (this.refs.view) {
+            this.refs.view.zoomOut().then(() => {
+                actions.selectPost('');
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: Why are we changing the selected post here?

File: app/screens/permalink/permalink.js [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+
+            if (this.refs.view) {
+                this.refs.view.zoomOut().then(() => {
+                    actions.selectPost('');
+
+                    if (onClose) {
+                        onClose();
+                    }
+
+                    navigator.resetTo({
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: I think you could use `navigator.popToRoot` here to skip having to redefine the root screen

File: app/screens/permalink/permalink.js [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+                            statusBarHideWithNavBar: false,
+                            screenBackgroundColor: theme.centerChannelBg
+                        }
+                    });
+
+                    if (channelTeamId && currentTeamId !== channelTeamId) {
+                        handleTeamChange(channelTeamId, false);
+                    }
+
+                    setChannelLoading(channelId !== currentChannelId);
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: We should come up with a way to have this baked into the channel/team change action automatically since we keep having to remember to call all of these wherever we change channels. Nothing to do now unless you have any good ideas for doing that

File: app/screens/permalink/permalink.js [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+                });
+            }
+        }
+    };
+
+    handlePress = () => {
+        const {channelId, channelName} = this.props;
+
+        if (this.refs.view) {
+            this.refs.view.growOut().then(() => {
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: It looks like we do these animations in two steps since we `growOut` here and then `zoomOut` above. Are both of those necessary, or should they be done at the same time?

File: app/screens/permalink/permalink.js [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+
+        const post = await actions.getPostThread(focusedPostId, false);
+        if (post.error && (!postIds || !postIds.length)) {
+            if (isPermalink && post.error.message.toLowerCase() !== 'network request failed') {
+                this.setState({
+                    error: formatMessage({
+                        id: 'permalink.error.access',
+                        defaultMessage: 'Permalink belongs to a deleted message or to a channel to which you do not have access.'
+                    }),
+                    title: formatMessage({
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: Instead of needing to track the title separately, couldn't we just have the render function show this when there's an error or just show the channel's name otherwise?

File: app/screens/permalink/permalink.js [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+                        id: 'mobile.search.no_results',
+                        defaultMessage: 'No Results Found'
+                    })
+                });
+                return;
+            }
+
+            this.setState({error: post.error.message, retry: true});
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: Can you move this into an else statement for the above if and just use a single return for both of them? This logic is hard to follow

File: app/screens/search/search.js [hunk 15]
```diff
@@ -218,11 +224,36 @@ class Search extends PureComponent {
...
-            focusedChannelId: post.channel_id,
-            focusedPostId: post.id
-        });
+        this.setState({isPermalink: false});
+        this.showPermalinkView(post.id);
+    };
+
+    showPermalinkView = (postId) => {
+        const {actions, navigator} = this.props;
+        const {isPermalink, showingPermalink} = this.state;
```
Comment (hmhealey) [inline app/screens/search/search.js]: Same comments as above about these two

File: app/screens/user_profile/user_profile.js [hunk 16]
```diff
@@ -47,17 +46,20 @@ class UserProfile extends PureComponent {
     }
 
     close = () => {
-        const {navigator} = this.props;
-
-        navigator.popToRoot({
-            animated: true
+        const {navigator, theme} = this.props;
+
+        navigator.resetTo({
```
Comment (hmhealey) [inline app/screens/user_profile/user_profile.js]: Why not use `popToRoot` here?

Comment (enahum) [reply app/screens/user_profile/user_profile.js]: `popToRoot` does not work well if you have one or multiple modals opened.

you can:
* Open search (modal)
* tap on a post (second modal)
* tap on a post (push to thread view)
* tap on a user (push to user profile view)
* Send message to user

File: app/screens/permalink/permalink.js (hunk 10 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: same as in the user profile

File: app/screens/permalink/permalink.js (hunk 11 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: please create a ticket to do that ;)

File: app/screens/permalink/permalink.js (hunk 9 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: I need to clear it, if not this gives an error with the PostTextBox in the center channel

File: app/screens/permalink/permalink.js (hunk 13 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: if this happens you may not have the channel name, that is why we are changing the title here

**Comment by enahum** [comment]
@hmhealey changes made

File: app/components/markdown/markdown_link/markdown_link.js [hunk 17]
```diff
@@ -12,36 +12,54 @@ import mattermostManaged from 'app/mattermost_managed';
...
+                if (supported) {
+                    Linking.openURL(url);
+                }
+            });
+            return;
+        }
+
+        const teamName = match[1];
+        const postId = match[2];
+        onPermalinkPress(postId, teamName);
```
Comment (hmhealey) [inline app/components/markdown/markdown_link/markdown_link.js]: I don't think https://github.com/mattermost/mattermost-mobile/pull/1449#discussion_r169663880 was addressed here

File: app/screens/permalink/permalink.js [hunk 18]
```diff
@@ -0,0 +1,455 @@
...
+                markChannelAsViewed,
+            } = actions;
+
+            actions.selectPost('');
+
+            if (onClose) {
+                onClose();
+            }
+
+            navigator.resetTo({
```
Comment (hmhealey) [inline app/screens/permalink/permalink.js]: In regards to https://github.com/mattermost/mattermost-mobile/pull/1449#discussion_r169676174, I think we should change back to popToRoot here if we can. There's no sense in redefining all this since it could easily be forgotten later on if we change how the root is defined.

File: app/screens/user_profile/user_profile.js (hunk 16 above)
Comment (hmhealey) [reply app/screens/user_profile/user_profile.js]: What happens with `popToRoot` in that case that's different from `resetTo`? It sounds like they should work the same by closing all the modals and screens to get back to the channel.

File: app/screens/permalink/permalink.js (hunk 18 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: It does not work with popToRoot try it out plz

File: app/components/markdown/markdown_link/markdown_link.js (hunk 17 above)
Comment (enahum) [reply app/components/markdown/markdown_link/markdown_link.js]: then I don't follow, the team name is passed down to the onPermalinkPress prop and then we load the channels for that team, what am I missing?

File: app/screens/permalink/permalink.js (hunk 18 above)
Comment (enahum) [reply app/screens/permalink/permalink.js]: and you may think that `popToRoot` is enough as I did as well, but when testing you see how it is not

File: app/components/markdown/markdown_link/markdown_link.js (hunk 17 above)
Comment (hmhealey) [reply app/components/markdown/markdown_link/markdown_link.js]: I mean that we could just pass down the permalink and have the post component (or whatever defines onPermalinkPress) call it. I'm 0/5 though now since I didn't realize that this is the only place that we were parsing the permalink.
